airflow/plugins/utils.py

```diff
@@ -81,7 +153,369 @@ def get_successfully_downloaded_feeds(execution_date):
+            **{PARTITIONED_ARTIFACT_METADATA_KEY: self.json()},
+        )
+
+
+def fetch_all_in_partition(
+    cls: typing.Type[PartitionedGCSArtifact],
+    fs: gcsfs.GCSFileSystem,
+    bucket: str,
+    table: str,
+    partitions: Dict[str, PartitionType],
+) -> List[typing.Type[PartitionedGCSArtifact]]:
+    path = "/".join(
+        [
+            bucket,
+            table,
+            *[f"{key}={value}" for key, value in partitions.items()],
+        ]
+    )
+    files = parse_obj_as(
+        GCSObjectInfoList,
+        [v for _, _, files in list(fs.walk(path, detail=True)) for v in files.values()],
+    )
+    return [
+        parse_obj_as(
+            cls, json.loads(fs.getxattr(file.name, PARTITIONED_ARTIFACT_METADATA_KEY))
+        )
+        for file in files.__root__
+    ]
+
+
+class ProcessingOutcome(BaseModel, abc.ABC):
+    success: bool
+    exception: Optional[Exception]
+    input_record: BaseModel
+
+    class Config:
+        arbitrary_types_allowed = True
+        json_encoders = {Exception: lambda e: str(e)}
+
+    # TODO: is this really useful?
+    @property
+    @abc.abstractmethod
+    def input_type(self) -> typing.Type[PartitionedGCSArtifact]:
+        """The input type that was processed to produce this outcome.."""
+
+
+class GCSBaseInfo(BaseModel, abc.ABC):
+    bucket: str
+    name: str
+
+    class Config:
+        extra = "ignore"
+
+    @property
+    @abc.abstractmethod
+    def partition(self) -> Dict[str, str]:
+        pass
+
+
+class GCSFileInfo(GCSBaseInfo):
+    type: Literal["file"]
+    size: int
+    md5Hash: str
+
+    @property
+    def filename(self) -> str:
+        return os.path.basename(self.name)
+
+    @property
+    def partition(self) -> Dict[str, str]:
+        return partition_map(self.name)
+
+
+class GCSObjectInfoList(BaseModel):
+    __root__: List["GCSObjectInfo"]
+
+
+class GCSDirectoryInfo(GCSBaseInfo):
+    type: Literal["directory"]
+
+    @property
+    def partition(self) -> Dict[str, str]:
+        return partition_map(self.name + "/")
+
+    def children(self, fs) -> List["GCSObjectInfo"]:
+        # TODO: this should work with a discriminated type but idk why it's not
+        return parse_obj_as(
+            GCSObjectInfoList, [fs.info(child) for child in fs.ls(self.name)]
+        ).__root__
+
+
+GCSObjectInfo = Annotated[
+    Union[GCSFileInfo, GCSDirectoryInfo],
+    Field(discriminator="type"),
+]
+
+GCSObjectInfoList.update_forward_refs()
+
+
+def get_latest_file(table_path: str, keys: List[str]) -> GCSFileInfo:
+    fs = get_fs()
+    directory = GCSDirectoryInfo(**fs.info(table_path))
+
+    for key in keys:
+        # TODO: this just sorts by string values, but we should parse stuff out
+        directory = sorted(
+            directory.children(fs), key=lambda o: o.partition[key], reverse=True
+        )[0]
+
+    children = directory.children(fs)
+    if len(children) != 1:
```

Comment:
clarification: is this an *us* thing or a general hive partitions thing to expect only one file in the final folder? are we sure this will always be the case?

Reply:
This is specific to us and I could be convinced otherwise. For example, maybe we just partition down to the e.g. hour and then timestamp within the underlying files.

Reply:
I think that this is ok but I wonder if we want to at least add a comment here being explicit about this just in case we do need to change it later